l2l/optimizers/kalmanfilter/enkf.py

```diff
@@ -0,0 +1,155 @@
+        :param  gamma: nd numpy array, Normalizes the model-data distance in the
+            update step, :`noise * I` (I is identity matrix) or
+            :math:`\\gamma=I` if `noise` is zero
+        :param p: nd numpy array
+            Exact solution given by :math:`G * u_exact`, where `G` is inverse
+            of a linear elliptic function `L`, it maps the control into the
+            observed data, Default is `None`
+        :return self, Possible outputs are:
+            ensembles: nd numpy array, optimized `ensembles`
+            Cpp: nd numpy array, covariance matrix of the model output
+            Cup: nd numpy array, covariance matrix of the model output and the
+                ensembles
+        """
+        model_output = model_output.copy()
+        # get shapes
+        self.gamma_s, self.dims = _get_shapes(observations, model_output)
+
+        if isinstance(gamma, (int, float)):
+            if float(gamma) == 0.:
+                self.gamma = np.eye(self.gamma_s)
+        else:
+            self.gamma = gamma
+
+        # copy the data so we do not overwrite the original arguments
+        self.ensemble = ensemble.copy()
+        self.observations = observations.copy()
+        self.observations = _encode_targets(observations, self.gamma_s)
+
+        for i in range(self.maxit):
+            # now get mini_batches
+            if self.n_batches > self.dims:
+                num_batches = 1
+            else:
+                num_batches = self.n_batches
+            mini_batches = _get_batches(
+                num_batches, shape=self.dims, online=self.online)
+
+            for idx in mini_batches:
+                # in case of online learning idx should be an int
+                # put it into a list to loop over it
+                if not isinstance(idx, np.ndarray):
+                    idx = [idx]
+
+                for d in idx:
+                    # now get only the individuals output according to idx
+                    g_tmp = model_output[:, :, d]
+                    # Calculate the covariances
+                    Cpp = _cov_mat(g_tmp, g_tmp, ensemble_size)
+                    Cup = _cov_mat(self.ensemble, g_tmp, ensemble_size)
+                    self.ensemble = _update_step(self.ensemble,
+                                                 self.observations[d],
+                                                 g_tmp, self.gamma,
+                                                 Cpp, Cup)
+        return self
+
+
+@jit(nopython=True, parallel=True)
+def _update_step(ensemble, observations, g, gamma, Cpp, Cup):
+    """
+    Update step of the kalman filter
+
+    Calculates the covariances and returns new ensembles
+    """
+    return ensemble + (
+                Cup @ np.linalg.lstsq(Cpp + gamma, (observations - g).T)[0]).T
+
+
+@jit(parallel=True)
+def _cov_mat(x, y, ensemble_size):
+    """
+    Covariance matrix
+    """
+    return np.tensordot((x - np.mean(x, axis=0)), (y - np.mean(y, axis=0)),
```

Comment:
Might be clearer if you use `np.einsum` here.

Reply:
For me it was the other away around. I had two reasons I used `tensordot`. The first one was I found it more intuitive than `einsum`. The equivalent formulation in `einsum` is:
```python
np.einsum('ij, il -> jl', x - np.mean(x, 0), y - np.mean(y, 0))
```  
For my understanding it not clear how the addition and dot product here are executed. In the tensordot formulation I know from the axes.

The second reason, probably the one more significant one, is the performance of `tensordot` over `einsum`. Unfortunately I did not do any timing besides just checking the time with the `time` module, einsum takes twice the time in my calculations.

Reply:
:+1: